logstash-core/src/main/java/org/logstash/config/ir/compiler/EventCondition.java

```diff
@@ -82,9 +82,13 @@ private Compiler() {
          * @return Compiled {@link EventCondition}
          */
         public static EventCondition buildCondition(final BooleanExpression expression) {
+            final String cachekey = expression.toRubyString();
+            EventCondition cached = CACHE.get(cachekey);
+            if (cached != null) {
+                return cached;
+            }
             synchronized (CACHE) {
```

Comment:
Much improved, but wouldn't it be cleaner / avoid contention to use a flow more like:

```java
EventCondiion buildCondition(final BooleanExpression expression) {
  final String cacheKey = expression.toRubyString();
  return CACHE.computeIfAbsent(EventCondition::buildConditionUncached)
}
```

AFAIK this prevents us from locking the whole datastructure, but rather, we just lock the key, which should increase liveness. Plus, it's easier to read.

Reply:
@andrewvc sorry I'm a little confused :), so this point is resolved, right?  (as you said I explained why we can't do this in the PR description)

Reply:
@andrewvc sorry only now realized there was a main-line comment.

Funny enough ... you're right the JIRA's seem to only address the recursive situation for `#computeIfAbsent` ... I'm able to reproduce the infinite loop without recursion here though, doing exactly what you suggest there. Will try to create a more minimal reproducer an open a JIRA in OpenJDK :)